cfs_config_util/update_components.py

```diff
@@ -0,0 +1,109 @@
+#
+# MIT License
+#
+# (C) Copyright 2023 Hewlett Packard Enterprise Development LP
+#
+# Permission is hereby granted, free of charge, to any person obtaining a
+# copy of this software and associated documentation files (the "Software"),
+# to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense,
+# and/or sell copies of the Software, and to permit persons to whom the
+# Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included
+# in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+"""
+Implementation of the update-components action of cfs-config-utility
+"""
+import logging
+
+from cfs_config_util.environment import (
+    API_CERT_VERIFY,
+    API_GW_HOST,
+)
+from cfs_config_util.errors import CFSConfigUtilError
+from cfs_config_util.hsm import get_node_ids
+from cfs_config_util.wait import wait_for_component_configuration
+
+from csm_api_client.service.cfs import (
+    CFSClient
+)
+from csm_api_client.service.gateway import APIError
+from csm_api_client.service.hsm import HSMClient
+from csm_api_client.session import AdminSession
+
+LOGGER = logging.getLogger(__name__)
+
+
+def update_cfs_components(cfs_client, component_ids, desired_config=None, clear_state=None,
+                          clear_error=None, enabled=None):
+    """Assign the CFSConfiguration to the given CFS components.
+
+    Args:
+        cfs_client (csm_api_client.service.cfs.CFSClient): the CFS API client
+        component_ids (Iterable): the list of component ids (xnames) to update
+        desired_config (str, Optional): the name of the desired config to set
+            on the components, if any
+        clear_state (bool, Optional): if True, clear state of the component
+        clear_error (bool, Optional): if True, clear errorCount of the component
+        enabled (bool, Optional): if specified, set the enabled of the component
+
+    Returns:
+        None
+
+    Raises:
+        CFSConfigUtilError: if unable to assign the CFS configuration to any
+            of the requested components.
+    """
+    failed_components = []
+    for component_id in component_ids:
+        try:
+            cfs_client.update_component(component_id, desired_config=desired_config,
```

Comment:
fyi, CFS does support a bulk patch endpoint, so you don't have to make these queries one at a time.

Reply:
Good to know. I'll have to look into that, and maybe I can update it as part of the CFS v2 -> v3 work.